### ISSUE TITLE
pyopenssl: maximum recursion depth exceeded while calling a Python object

### ISSUE THREAD

**Issue by gsakkis** [urllib3/urllib3#681]
I opened this issue on [requests](https://github.com/kennethreitz/requests/issues/2681) but just realized it's actually in the vendored urllib3.
##

**Comment by Lukasa** [comment]
Bringing the description over from requests:

I recently installed requests[security] to get rid of the InsecurePlatformWarning and have been getting intermittent Runtime errors in pyopenssl. Here's a sample stacktrace:

```
Stacktrace (most recent call last):

  File "requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "requests/sessions.py", line 605, in send
    r.content
  File "requests/models.py", line 750, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "requests/models.py", line 673, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "requests/packages/urllib3/response.py", line 307, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "requests/packages/urllib3/response.py", line 243, in read
    data = self._fp.read(amt)
  File "python2.7/httplib.py", line 567, in read
    s = self.fp.read(amt)
  File "python2.7/socket.py", line 380, in read
    data = self._sock.recv(left)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    return self.recv(*args, **kwargs)
  File "requests/packages/urllib3/contrib/pyopenssl.py", line 171, in recv
    data = self.connection.recv(*args, **kwargs)
  File "OpenSSL/SSL.py", line 1318, in recv
    buf = _ffi.new("char[]", bufsiz)
  File "cffi/api.py", line 235, in new
    if isinstance(cdecl, basestring):
```

Any ideas?

**Comment by Lukasa** [comment]
I'd like to see the full stacktrace, because it got chopped at the bottom, but this *looks* like a recursion depth error to me.

**Comment by shazow** [comment]
I guess it's this line? https://github.com/shazow/urllib3/blob/master/urllib3/contrib/pyopenssl.py#L196

Is recursing here a good idea to begin with, rather than raising the error?

**Comment by Lukasa** [comment]
Recursing here is expected behaviour. You *might* want to refactor the recursion to iteration, but the retry logic is wrong: `WantReadError` is an instruction from OpenSSL to read more data.

**Comment by shazow** [comment]
Mmm control flow with exceptions, yum.

So is this reproducible with a sufficiently large/slow payload?

**Comment by Lukasa** [comment]
I don't think so: it's something to do with the TLS, actually. It essentially means that there's TLS data that still needs to be received before the payload can be sent/received. It's weird.

**Comment by shazow** [comment]
Hmmm any idea how to reproduce this reliably? I mean, would it be safe to convert to iterative, or would it just loop infinitely anyways?

**Comment by Lukasa** [comment]
Not off the top of my head, but @t-8ch might. As to looping forever, I think you definitely need a cutoff point somewhere.

**Comment by t-8ch** [comment]
I also think we do the right thing here (minus the recursion).
Do we want to perform our own timeout handling?

@gsakkis Did you test wether this issue has already been fixed in urllib3?
To debug this I would greatly appreciate the following information:
* Version of requests/urllib3
* Exact parameters used for your requests call
* If possible a publicly available URL to test against

**Comment by gsakkis** [comment]
@t-8ch no I haven't tested it with the urllib3 directly, only with the latest `requests[security]`. 
- Installed package versions:
  - cffi==1.1.2
  - cryptography==0.9.3
  - idna==2.0
  - ipaddress==1.0.14
  - ndg-httpsclient==0.4.0
  - pyOpenSSL==0.15.1
  - pyasn1==0.1.8
  - pycparser==2.14
  - requests==2.7.0

- The only non-default parameters sent to requests are a `User-Agent` header , `timeout=10`, and `proxies= {'http': 'http://localhost:5555', 'https': 'http://localhost:5555'}` where this is a local HAProxy server.
- I could provide a few public urls but they probably wouldn't be very useful. The failures are both non-reproducible and rare; just a hundred or two out of tens/hundreds of thousand successful ones.

**Comment by carlopires** [comment]
I'm facing this error in an intermittent fashion. Using:
- requests (2.10.0)
- pyOpenSSL (16.0.0)
- urllib3 (1.15.1)
- ndg-httpsclient (0.4.1) 

Do we have a fix for this?

**Comment by sigmavirus24** [comment]
@carlopires how did you install those?

**Comment by carlopires** [comment]
@sigmavirus24 I have a requirements.txt file with lines like:
- pyOpenSSL>=16.0.0
- requests[security]==2.10.0
- urllib3>=1.15.1

Then I install using 
    
    pip install -r requirements.txt

**Comment by sigmavirus24** [comment]
@carlopires and on what operating system with what version of openssl? (Just to have all the facts)

**Comment by carlopires** [comment]
ubuntu@carlo2:~$ cat /etc/issue
    Ubuntu 14.04.4 LTS \n \l
    
    ubuntu@carlo2:~$ openssl version
    OpenSSL 1.0.1f 6 Jan 2014

**Comment by Lukasa** [comment]
Are you encountering this error with a website that's publicly reachable?

**Comment by carlopires** [comment]
It is happening on several sites. Two examples:
- https://outlook.office.com/api
- https://_customer_domain_.sf-api.com/sf/v3 (Citrix ShareFile API)

**Comment by Lukasa** [comment]
Hrrm. Right now I'm not able to reproduce this myself. Does it happen consistently for you or intermittently?

**Comment by carlopires** [comment]
It always happen in my test suite.. which run thousands of https calls to lots of sites. But the behavior is intermittent for sure.. it appears in different https calls.

**Comment by Lukasa** [comment]
@carlopires In that case, it would be *extremely* valuable if you were able to get hold of a .pcap file showing the specific request that is failing.

Can you also confirm what the actual exception is? Is it a max recursion depth error?

**Comment by xelhark** [comment]
Any news about this? I have the same behavior and I would really like to remove these errors from my reports..

**Comment by Lukasa** [comment]
@xelhark We're still a little bit surprised that this is hittable at all. Do you have any way to get a packet capture for this?

**Comment by xelhark** [comment]
@Lukasa

Sorry for the late answer.

We are experiencing this issue regularly, but it's happening on my remote server so I'm not sure I can capture packages.

Here's the code where this happens:

` def recv(self, *args, **kwargs):
        try:
            data = self.connection.recv(*args, **kwargs)
        except OpenSSL.SSL.SysCallError as e:
            if self.suppress_ragged_eofs and e.args == (-1, 'Unexpected EOF'):
                return b''
            else:
                raise SocketError(str(e))
        except OpenSSL.SSL.ZeroReturnError as e:
            if self.connection.get_shutdown() == OpenSSL.SSL.RECEIVED_SHUTDOWN:
                return b''
            else:
                raise
        except OpenSSL.SSL.WantReadError:
            rd, wd, ed = select.select(
                [self.socket], [], [], self.socket.gettimeout())
            if not rd:
                raise timeout('The read operation timed out')
            else:
                return self.recv(*args, **kwargs) # Max Recursion Depth error here -----
        else:
            return data`

Please let me know if there's any data I can provide that can help you with this. I can't reproduce it systematically unfortunately, but it happens regularly, I should be able to grab some data, just tell me what do you need and how can I get it.

**Comment by Lukasa** [comment]
I'm entirely familiar with the code and where this is happening. What's not clear to me, *at all*, is how we can be spinning in recv() to the point that we exceed the recursion depth. Essentially, the socket is readable, but somehow OpenSSL is concluding that there isn't enough data there. I don't understand how that can be yet.

**Comment by xelhark** [comment]
Is there any data I can dig to help?

I'm using Sentry to monitor the errors, I have access to the stack trace and `__repr__` of all data while executing that code. Not sure if this helps at all:

```
            rd, wd, ed = select.select(
                [self.socket], [], [], self.socket.gettimeout())
            if not rd:
                raise timeout('The read operation timed out')
            else:
                return self.recv(*args, **kwargs)
```

```
args: [
    7297
]

ed: []

kwargs: {}

rd:[
    <eventlet.greenio.base.GreenSocket object at 0x7fdafbbcdb10>
]

self: 
<requests.packages.urllib3.contrib.pyopenssl.WrappedSocket object at 0x7fdafbbf8910>

wd:
[]
```

**Comment by Lukasa** [comment]
Hrm, that's a little bit interesting. What *hadn't* been made clear to me is that you're using eventlet. I wonder if there's a problem here with the fact that eventlet puts the sockets into non-blocking mode. The fact that we select over the socket suggests it shouldn't be an issue, but it's possible I suppose.

Regardless, that data really isn't enough. We would need to see what the actual traffic flow is, I suspect. Does eventlet patch our PyOpenSSL import?

**Comment by xelhark** [comment]
>Does eventlet patch our PyOpenSSL import?

I don't know. I found [this](http://eventlet.net/doc/changelog.html#id1) in the changelog. (I'm using 0.19.0).

Our code does not contain any monkey patch though.

**Comment by Lukasa** [comment]
No monkey patch at all? Of anything?

**Comment by xelhark** [comment]
Not on our side, but, probably eventlet does?

http://eventlet.net/doc/patching.html

I'm not sure, I'm no expert at this.

**Comment by Lukasa** [comment]
Either you or one of your dependencies must be patching the standard library, because our socket is an eventlet one, not a standard library one.

**Comment by xelhark** [comment]
Yes, I think eventlet does:

http://eventlet.net/doc/patching.html

**Comment by Lukasa** [comment]
Eventlet doesn't *automatically* patch the standard library, as far as I'm aware: someone has to ask it to. Given that your application depends on eventlet, it would be safe to assume that your application does that. Can you search through your codebase quickly for the string `eventlet` to confirm?

**Comment by gsakkis** [comment]
For the record we're not using Eventlet. 

Btw this has stopped appearing in our logs since November though I can't find any potentially related changes in git log around these dates, so no idea really of what changed.

**Comment by Lukasa** [comment]
@gsakkis *Were* you using eventlet? Just asking out of blind hope more than anything else. =P

**Comment by gsakkis** [comment]
@Lukasa no, never; sorry :)

**Comment by sigmavirus24** [comment]
@gsakkis perhaps your dependency list has changed since https://github.com/shazow/urllib3/issues/681#issuecomment-123653877 ?

**Comment by gsakkis** [comment]
@sigmavirus24 it did change but only after a few months since the last appearance of this stacktrace in the logs. I can paste the current dependency versions if you think it might be relevant though.

**Comment by carlopires** [comment]
Here is how I fixed this issue (I'm using eventlet + requests):

```
def eventleted_recv(self, *args, **kwargs):
    time_limit = self.socket.gettimeout()
    time_start = time.time()
    while 1:
        try:
            data = self.connection.recv(*args, **kwargs)
        except OpenSSL.SSL.SysCallError as e:
            if self.suppress_ragged_eofs and e.args == (-1, 'Unexpected EOF'):
                return b''
            else:
                raise SocketError(str(e))
        except OpenSSL.SSL.ZeroReturnError as e:
            if self.connection.get_shutdown() == OpenSSL.SSL.RECEIVED_SHUTDOWN:
                return b''
            else:
                raise
        except OpenSSL.SSL.WantReadError:
            eventlet.sleep(0)

            if time_limit is None:
                continue

            time_elapsed = time.time() - time_start
            time_limit = max(0, time_limit - time_elapsed)

            rd, _wd, _ed = select.select([self.socket], [], [], time_limit)
            if not rd:
                raise TimeoutError('The read operation timed out')
        else:
            return data

from requests.packages.urllib3.contrib.pyopenssl import WrappedSocket
WrappedSocket.recv = eventleted_recv
```

I had to make similar like patches to ```WrappedSocket.recv_into``` and ```WrappedSocket._send_until_done``` too.

The bug is related to how *WrappedSocket* deals with *WantReadError*.

**Comment by Lukasa** [comment]
Well, that has alerted me to one probable misbehaviour: `select` is being called with `self.socket.gettimeout()`, which on eventlet sockets is almost certainly set to `0`. That would cause select to return immediately. However, unless the socket is readable this should not recurse: instead it should raise a socket timeout. So my question is: why isn't it?

**Comment by sethmlarson** [comment]
Since there has been no new reports of this issue in some time, I'm going to close this as resolved / wontfix? We can open a new issue if it arises again.

**Comment by fabiocaccamo** [comment]
@sethmlarson I'm experiencing this issue on a project running on:

- Python 3.11
- Gunicorn 21.2.0
- Django 4.2.3
- urllib3 1.26.16 *(I can't use the latest version for compatibility with other installed packages)*

**In my case the problem was related to the gunicorn `--preload` option, removing that option fixed the issue.**

I hope that my feedback could be useful to fix this issue.

**Comment by SuperWhite18** [comment]
> @sethmlarson I'm experiencing a similar issue on a project running on:
> 
> * Python 3.11
> * Gunicorn 21.2.0
> * Django 4.2.3
> * urllib3 1.26.16 _(I can't use the latest version for compatibility with other installed packages)_
> 
> Here the error traceback:
> 
> ```python
> File "/root/myproject/src/accounts/auth.py", line 63, in _email_is_disposable
>     response = requests.get(url)
>                ^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/requests/api.py", line 73, in get
>     return request("get", url, params=params, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/requests/api.py", line 59, in request
>     return session.request(method=method, url=url, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
>     resp = self.send(prep, **send_kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
>     r = adapter.send(request, **kwargs)
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/requests/adapters.py", line 486, in send
>     resp = conn.urlopen(
>            ^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 703, in urlopen
>     httplib_response = self._make_request(
>                        ^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 386, in _make_request
>     self._validate_conn(conn)
>   File "/root/myproject/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 1042, in _validate_conn
>     conn.connect()
>   File "/root/myproject/venv/lib/python3.11/site-packages/urllib3/connection.py", line 400, in connect
>     self.ssl_context = create_urllib3_context(
>                        ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/myproject/venv/lib/python3.11/site-packages/urllib3/util/ssl_.py", line 312, in create_urllib3_context
>     context.options |= options
>     ^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/ssl.py", line 624, in options
>     super(SSLContext, SSLContext).options.__set__(self, value)
>   File "/usr/lib/python3.11/ssl.py", line 624, in options
>     super(SSLContext, SSLContext).options.__set__(self, value)
>   File "/usr/lib/python3.11/ssl.py", line 624, in options
>     super(SSLContext, SSLContext).options.__set__(self, value)
>   [Previous line repeated 447 more times]
> RecursionError: maximum recursion depth exceeded while calling a Python object
> ```
> 
> ### In my case the problem was related to the gunicorn `--preload` option, removing that option fixed the issue.
> I hope that my feedback will be useful.

Please tell me u how to solve this problem, I have a same question as following, thanks a lot !!:

> `Traceback (most recent call last):
  File "/home/root/robot/src/modules/judgment/survey.py"", line 278, in have_tianqing
    status_code, text = TQiApi().eye_client_info(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/robot/src/modules/judgment/survey.py", line 97, in eye_client_info
    res = requests.post(url=url, headers=self.headers, json=json_, verify=False)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/requests/api.py", line 115, in post
    return request("post", url, data=data, json=json, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 467, in _make_request
    self._validate_conn(conn)
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 1099, in _validate_conn
    conn.connect()
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/connection.py", line 653, in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/connection.py", line 759, in _ssl_wrap_socket_and_match_hostname
    context = create_urllib3_context(
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/root/venv/lib/python3.11/site-packages/urllib3/util/ssl_.py", line 290, in create_urllib3_context
    context.minimum_version = TLSVersion.TLSv1_2
    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/ssl.py", line 608, in minimum_version
    super(SSLContext, SSLContext).minimum_version.__set__(self, value)
  File "/usr/lib/python3.11/ssl.py", line 608, in minimum_version
    super(SSLContext, SSLContext).minimum_version.__set__(self, value)
  File "/usr/lib/python3.11/ssl.py", line 608, in minimum_version
    super(SSLContext, SSLContext).minimum_version.__set__(self, value)
  [Previous line repeated 482 more times]
  File "/usr/lib/python3.11/ssl.py", line 606, in minimum_version
    if value == TLSVersion.SSLv3:
                ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/enum.py", line 193, in __get__
    def __get__(self, instance, ownerclass=None):
    
  File "/usr/lib/python3.11/enum.py", line 193, in __get__
    def __get__(self, instance, ownerclass=None):
    
  File "_pydevd_bundle/pydevd_cython.pyx", line 1457, in _pydevd_bundle.pydevd_cython.SafeCallWrapper.__call__
RecursionError: maximum recursion depth exceeded while calling a Python object`

**Comment by dbarnett** [comment]
> Since there has been no new reports of this issue in some time, I'm going to close this as resolved / wontfix? We can open a new issue if it arises again.

Not resolved. Could you reopen and help narrow down a fix/workaround for all of us still affected? Still very much seeing this: https://github.com/insanum/gcalcli/issues/735.

It's really frustrating because it seems to come from some combination of certain dependencies but I have no idea which project tracker to follow up in. Hopefully this one is the right place. From https://stackoverflow.com/questions/49820173/recursionerror-maximum-recursion-depth-exceeded-from-ssl-py-supersslcontex, it looks like some monkey patching garbage in gevent and similar libraries, but in my case I can't find anything installed in the environment that might be doing such weird monkey patching:
```shell
$ pip list
Package                        Version              Editable project location
------------------------------ -------------------- ----------------------------
argcomplete                    3.5.0
cachetools                     5.5.0
certifi                        2024.7.4
charset-normalizer             3.3.2
coverage                       7.6.1
dateparser                     1.2.0
flake8                         7.1.1
gcalcli                        4.4.1.dev20+g610499f /home/david/projects/gcalcli
google-api-core                2.19.1
google-api-python-client       2.142.0
google-api-python-client-stubs 1.27.0
google-auth                    2.34.0
google-auth-httplib2           0.2.0
google-auth-oauthlib           1.2.1
googleapis-common-protos       1.63.2
httplib2                       0.22.0
idna                           3.7
iniconfig                      2.0.0
mccabe                         0.7.0
oauth2client                   4.1.3
oauthlib                       3.2.2
packaging                      24.1
parsedatetime                  2.6
pip                            24.0
pluggy                         1.5.0
proto-plus                     1.24.0
protobuf                       5.27.3
pyasn1                         0.6.0
pyasn1_modules                 0.4.0
pycodestyle                    2.12.1
pyflakes                       3.2.0
PyICU                          2.13.1
pyparsing                      3.1.2
pytest                         8.3.2
pytest-cov                     5.0.0
python-dateutil                2.9.0.post0
pytz                           2024.1
regex                          2024.7.24
requests                       2.32.3
requests-oauthlib              2.0.0
rsa                            4.9
six                            1.16.0
truststore                     0.9.2
types-httplib2                 0.22.0.20240310
typing_extensions              4.12.2
tzlocal                        5.2
uritemplate                    4.1.1
urllib3                        2.2.2
vobject                        0.9.7
```
